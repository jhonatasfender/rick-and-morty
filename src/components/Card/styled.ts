import styled, { css } from 'styled-components'

export const Wrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    flex-direction: column;
    gap: 1rem;
    overflow: hidden;
    border-radius: 0.8rem;
    width: 16rem;
    height: 30rem;
    border: 0.1rem solid ${theme.colors.white};
    cursor: pointer;
  `}
`

export const Image = styled.img`
  min-width: 16rem;
  width: 16rem;
`

export const Name = styled.span`
  ${({ theme }) => css`
    color: ${theme.colors.white};
    font-size: 1.9rem;
    display: block;
    width: 13rem;
    white-space: nowrap;
    text-overflow: ellipsis;
    overflow: hidden;
  `}
`

export const ContentCard = styled.div`
  padding: 1rem;
  display: flex;
  flex-direction: column;
  gap: 1rem;
`

export const Tags = styled.div`
  display: flex;
  justify-content: space-between;
  flex-wrap: wrap;
  gap: 1rem;
`

export const Tag = styled.div`
  ${({ theme }) => css`
    font-size: 1rem;
    font-weight: bold;
    background-color: ${theme.colors.hacienda};
    color: ${theme.colors.white};
    padding: 0.4rem 1rem;
    border-radius: 1rem;
    min-width: 45%;
    white-space: nowrap;
    text-overflow: ellipsis;
    overflow: hidden;
  `}
`
