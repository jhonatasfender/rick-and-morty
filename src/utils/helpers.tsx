import { ReactNode } from 'react';
import { QueryClient, QueryClientProvider } from 'react-query';
import { MemoryRouter } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';

import { render, RenderResult } from '@testing-library/react'

import theme from '../styles/theme'

const queryClient = new QueryClient()

export const renderWithThemeAndRoute = (children: ReactNode): RenderResult =>
  render(
    <QueryClientProvider client={queryClient}>
      <MemoryRouter>
        <ThemeProvider theme={theme}>{children}</ThemeProvider>
      </MemoryRouter>
    </QueryClientProvider>,
  )
