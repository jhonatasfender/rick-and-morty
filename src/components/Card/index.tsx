import { useNavigate } from 'react-router-dom'
import { Result } from '../../hooks/use-list/typings'
import * as S from './styled'

type CardProps = {
  result: Result
}

export default function Card({ result }: CardProps): JSX.Element {
  const navigate = useNavigate()

  const handleClick = () => {
    navigate(`/character/${result.id}`)
  }

  return (
    <S.Wrapper onClick={handleClick}>
      <S.Image src={result.image} />

      <S.ContentCard>
        <S.Name>{result.name}</S.Name>

        <S.Tags>
          <S.Tag>{result.species}</S.Tag>
          <S.Tag>{result.gender}</S.Tag>
          <S.Tag>{result.origin.name}</S.Tag>
          {result.type && <S.Tag>{result.type}</S.Tag>}
        </S.Tags>
      </S.ContentCard>
    </S.Wrapper>
  )
}
