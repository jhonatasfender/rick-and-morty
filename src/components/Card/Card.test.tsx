import { screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { renderWithThemeAndRoute } from '../../utils/helpers'
import Card from './'

const mockedUsedNavigate = jest.fn()

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as any),
  useNavigate: () => mockedUsedNavigate,
}))

const mock = {
  id: 14,
  name: 'Alien Morty',
  status: 'unknown',
  species: 'Alien',
  type: '',
  gender: 'Male',
  origin: { name: 'unknown', url: '' },
  location: {
    name: 'Citadel of Ricks',
    url: 'https://rickandmortyapi.com/api/location/3',
  },
  image: 'https://rickandmortyapi.com/api/character/avatar/14.jpeg',
  episode: ['https://rickandmortyapi.com/api/episode/10'],
  url: 'https://rickandmortyapi.com/api/character/14',
  created: '2017-11-04T20:51:31.373Z',
}

describe('<Card />', () => {
  it('should load the component with default values', () => {
    const { container } = renderWithThemeAndRoute(<Card result={mock} />)

    expect(container).toMatchSnapshot()
  })

  it('should load the type attribute', () => {
    renderWithThemeAndRoute(<Card result={{ ...mock, type: 'unique-type' }} />)

    expect(screen.getByText('unique-type')).toBeInTheDocument()
  })

  it('should redirect when click on card', () => {
    renderWithThemeAndRoute(<Card result={mock} />)

    userEvent.click(screen.getByRole('listitem'))

    expect(mockedUsedNavigate).toHaveBeenCalledWith('/character/14')
  })
})
