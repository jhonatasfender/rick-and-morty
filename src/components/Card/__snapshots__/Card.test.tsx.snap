// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> should load the component with default values 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 1rem;
  overflow: hidden;
  border-radius: 0.8rem;
  width: 16rem;
  height: 30rem;
  border: 0.1rem solid #ffffff;
  cursor: pointer;
}

.c1 {
  min-width: 16rem;
  width: 16rem;
}

.c3 {
  color: #ffffff;
  font-size: 1.9rem;
  display: block;
  width: 13rem;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
}

.c2 {
  padding: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 1rem;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  gap: 1rem;
}

.c5 {
  font-size: 1rem;
  font-weight: bold;
  background-color: #947F18;
  color: #ffffff;
  padding: 0.4rem 1rem;
  border-radius: 1rem;
  min-width: 45%;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
}

<div>
  <div
    class="c0"
    role="listitem"
  >
    <img
      class="c1"
      src="https://rickandmortyapi.com/api/character/avatar/14.jpeg"
    />
    <div
      class="c2"
    >
      <span
        class="c3"
      >
        Alien Morty
      </span>
      <div
        class="c4"
      >
        <div
          class="c5"
        >
          Alien
        </div>
        <div
          class="c5"
        >
          Male
        </div>
        <div
          class="c5"
        >
          unknown
        </div>
      </div>
    </div>
  </div>
</div>
`;
